### ISSUE TITLE
Feature request: Checks that `unless` is not used with multiple conditions.

### ISSUE THREAD

**Issue by wata727** [rubocop-hq/rubocop#5388]
I was worried about the poor readability of code using `unless` with multiple conditions.

```ruby
unless foo && bar
  something
end
```

Looking at the code like the above, I first think about De Morgan's law, rewrite it and proceed with implementation. This is a very painful task.

In order to prevent this problem, I created a cop called `Lint/UnlessMultipleCondtions`.
https://github.com/wata727/rubocop/commit/da4ed994d5fd263c4e08d4df98ea6cd716244ca3

However, when this Cop is executed to RuboCop, the result is as follows:

```
...

lib/rubocop/config_loader.rb:128:23: W: Lint/UnlessMultipleConditions: Avoid using unless with multiple conditions.
        return unless hash['AllCops'] && hash['AllCops'][version]
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1101 files inspected, 140 offenses detected
RuboCop failed!
```

I don't mind fixing these problems, but I doubt if this fix will be accepted by the community.

Please tell me your opinion :)

**Comment by mikegee** [comment]
This seems like a good idea. 140 offenses in Rubocop's code is a bad sign though. Maybe it will be too noisy and should be disabled by default?

**Comment by asherkach** [comment]
Disagree, I think the `unless foo && bar` structure can be far more readable than `if !foo && !bar` depending on what `foo` and `bar` are. Granted it can be far less readable for the cases where it is  natural to invoke De Morgan's Law. Banning the usage of conjunctions in `unless` statement seems like a bad idea to me.

**Comment by pocke** [comment]
It's a very nice cop!
BTW, I think the cop should be in the Style department instead of the Lint department. Because `unless foo && bar` style is not a bug, but the style confuses us and sometimes we make a bug with the style. And I think the cop should be enabled by default if it's in the Style department.

**Comment by mikegee** [comment]
> Disagree, I think the unless foo && bar structure can be far more readable than if !foo && !bar

That is true. When I said I thought this cop was a good idea, I was thinking developers should extract these more complicated conditions to a method.

**Comment by wata727** [comment]
While looking the code where the offenses were found, I noticed one thing.
In an environment where `Hash#dig` can not be used, the following code is often used for checking elements of a nested hash:

```ruby
return unless hash[:a] && hash[:a][:b]
```

I think that this code should not be fixed as follows.

```ruby
return if !hash[:a] || !hash[:a][:b]
```

In such a case, I think that the first example can be used as a kind of idiom. So, I'm wondering whether this cop should be enabled by default...

**Comment by Drenmi** [comment]
```ruby
cat.pet unless cat.lion? || cat.tiger?

cat.pet unless cat.large? && cat.fangs?
```

These examples both seem perfectly readable.

As per @mikegee's suggestion, one can implement `Cat#dangerous?`, but the question is if this generalizes across common examples.

**Comment by wata727** [comment]
> These examples both seem perfectly readable.

I agree with it. I think that there are occasionally bad examples, but I understood that there are circumstances to use like this.

> but the question is if this generalizes across common examples. 🤔

In order to avoid this Cop, I think that rewriting of understandable code should be avoided.

Based on these, I'm thinking whether to take either of the following:

1. Add this Cop with `Enabled: false`
2. Do not add this Cop

wdyt?

**Comment by wata727** [comment]
I will close the pull request once. Thank you very much for giving us your opinion.

**Comment by Morozzzko** [comment]
I've been hoping to see this cop for a long time. It's a shame it wasn't merged. 

Have you had any chance to revisit this issue and, perhaps, think about better impementation?

While I agree that _not all_ composite conditions are bad, especially in the multiline approach, I'd love to have an option to warn about conditions that look `return xxx unless a && b || c`

There's rarely a good reason to use such complex examples, to be honest.

I've found that I write this code for two reasons:

**Compaction**. If I have multiple guards, I can turn them into one

```
return unless user
return unless order.shipped?

# turns into

return unless user && order.shipped?
```

I still believe that this code is harder to reason about than the alternative. While the compaction may be good, it's still not as good as the composite if

```
return if user.nil? || !order.shipped?
```

or even better – some kind of predicate that encapsulates the logic

```
return unless applicable?
```

**Avoiding explicit `#nil?` checks**. While it will definitely trick `reek` into not firing [NilCheck](https://github.com/troessner/reek/blob/master/docs/Nil-Check.md) violation, it's still not a decent reason to make the code more complicated


While I believe that this cop should not be auto correctable, it should probably exist in _some way_

**Comment by MikeMcQuaid** [comment]
> I've been hoping to see this cop for a long time. It's a shame it wasn't merged.

Agreed!

> ```ruby
> cat.pet unless cat.lion? || cat.tiger?
> 
> cat.pet unless cat.large? && cat.fangs?
> ```
> 
> These examples both seem perfectly readable.

Just to offer my own $0.02: I've been working with Ruby professionally for ~8 years, through being a Homebrew maintainer for ~12 years and programming professionally (not just using Ruby) for ~14 years and: I find all cases of `unless ... ||` or `unless ... &&` (including the above) hard to mentally parse and, particularly, to accurately and consistent reverse the logic (by turning it into an `if`). 

That's not to say it should be a default cop or whether it should be a style/lint/whatever cop but: I think there is a valid need for a cop like this and it can improve readability (particularly for those coming from languages without an `unless` operator).

**Comment by Drenmi** [comment]
It seems there's good reason to think this cop would be useful to people, and the issue was only automatically closed by GitHub when the first PR was closed by it's author, so let's re-open this one. 🙂

**Comment by caalberts** [comment]
> While I agree that not all composite conditions are bad, especially in the multiline approach, I'd love to have an option to warn about conditions that look return xxx unless a && b || c

This is a good idea. The starting point could be to warn against using opposing boolean operators in `unless`.

```rb
# ok
return xxx unless a && b && c
return xxx unless a || b || c

# not ok
return xxx unless a && b || c ...
return xxx unless a || b && c ...
```

What should it do if the condition is within `()`, which would improve readability significantly?

```rb
return xxx unless a && (b || c)
```

**Comment by caalberts** [comment]
I created https://github.com/rubocop-hq/rubocop/pull/9386 to check for mixed use of `&&` and `||` in unless. I think this is a more lenient approach as compared to using unless with multiple conditions. Currently there are 6 offenses in Rubocop.

**Comment by Incanus3** [comment]
Hi,
I just updated rubocop to 1.11.0 and started getting

```ruby
C: Style/UnlessLogicalOperators: Do not use mixed logical operators in an unless.
```

on a line that reads

```ruby
unless values[:client][:insurance_company] == 999
```

1) I couldn't find any documentation for this cop, which is understandable if it's very new. I tend to set EnabledByDefault to true and then manually disable cops I don't like. If I didn't set this, the cop would probably be disabled.
2) I don't think there's a logical operator in this condition (equality operator is a boolean operator, but not a logical one AFAIK) and they definitely aren't mixed, so either there's a bug or I got the message completely wrong.

Could I get some info about this? If it is a bug, can I provide some more context? If not, what change am I supposed to make to fix this?

P.S.: Thanks for this awesome gem, I use it everywhere and it definitely makes my codebases better.

**Comment by bbatsov** [comment]
> Could I get some info about this? If it is a bug, can I provide some more context? If not, what change am I supposed to make to fix this?

It's a bug, so it'd be best to file a ticket about it. 

The docs are here - https://docs.rubocop.org/rubocop/1.11/cops_style.html#styleunlesslogicaloperators

**Comment by Incanus3** [comment]
Wow, that was fast.

> It's a bug, so it'd be best to file a ticket about it.

ok, will do.

> The docs are here - https://docs.rubocop.org/rubocop/1.11/cops_style.html#styleunlesslogicaloperators

hmm, that's strange, I looked like 3 mins before posting (without the version in the url) and couldn't find i there. now i just refreshed, used the same search string and immediately found it. thanks.

**Comment by Incanus3** [comment]
> > It's a bug, so it'd be best to file a ticket about it.
> 
> ok, will do.

[here](https://github.com/rubocop/rubocop/issues/9551)
